cmd/soapboxd/main.go

```diff
@@ -118,3 +124,55 @@ func timingInterceptor(
 	log.Printf("method=%s duration=%s error=%v", info.FullMethod, time.Since(t0), err)
 	return resp, err
 }
+
+func loginInterceptor(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	switch strings.Split(info.FullMethod, "/")[2] {
+	case "LoginUser", "CreateUser", "GetUser":
+		return handler(ctx, req)
+	default:
+		if err := authorize(ctx); err != nil {
+			return nil, err
+		}
+		return handler(ctx, req)
+	}
+}
+
+type accessDeniedErr struct {
+	userID []byte
+}
+
+func (e *accessDeniedErr) Error() string {
+	return fmt.Sprintf("Incorrect login token for user %s", e.userID)
+}
+
+type emptyMetadataErr struct{}
+
+func (e *emptyMetadataErr) Error() string {
+	return fmt.Sprint("No metadata attached with request")
+}
+
+func authorize(ctx context.Context) error {
+	if md, ok := metadata.FromIncomingContext(ctx); ok {
+		userID := []byte(md["user_id"][0])
+		sentToken, err := base64.StdEncoding.DecodeString(md["login_token"][0])
+		if err != nil {
+			return err
+		}
+		key := []byte(os.Getenv("LOGIN_SECRET_KEY"))
+		h := hmac.New(sha512.New, key)
+		h.Write(userID)
```

Comment:
Yeah, this doesn't seem great. Stealing this token would let you impersonate the user literally forever. I'll throw in a TODO.

Adding some randomness to the token isn't hard, but if we do that, we have to store it. So perhaps in the future we should put it in the database, tied to the user, along with a timestamp so it will expire automatically?

AFAIK soapboxd doesn't really have the concept of a session yet, so I'm very open to suggestions here.

Reply:
:+1:

The revocable bit is probably more important (since that's how most API tokens work for command-line etc.)

The session scoping would only make sense if we got rid of the CLI and went web-app only since then you could do an Oauth flow to get a JWT and present that.

Some authentication is better than none so may as well ship this and revisit.